File: cerebras/gtest/test_heaviside.cpp
// test_heaviside.cpp
#include <gtest/gtest.h>
#include <torch/torch.h>
#include "heaviside.h"

namespace custom_namespace {

    // Test for positive inputs to compute_heaviside
    TEST(HeavisideTest, PositiveInput) {
        custom_namespace::Heaviside heaviside_instance;
        torch::Tensor input = torch::tensor({1.0, 2.0, 3.0});
        torch::Tensor values = torch::tensor({-1.0, -2.0, -3.0});
        
        // Call compute_heaviside
        torch::Tensor result = heaviside_instance.compute_heaviside(input, values);

        // Expected output: tensor of ones since input is positive
        torch::Tensor expected = torch::heaviside(input, values);

        ASSERT_TRUE(torch::allclose(result, expected, 1e-4));
    }

    // Test for zero input with default h value
    TEST(HeavisideTest, ZeroInputDefaultH) {
        custom_namespace::Heaviside heaviside_instance;
        torch::Tensor input = torch::tensor({0.0});
        torch::Tensor values = torch::tensor({0.0});
        
        // Call compute_heaviside with default h (0.5)
        torch::Tensor result = heaviside_instance.compute_heaviside(input, values);

        // Expected output: tensor of 0.5 for input of zero
        torch::Tensor expected = torch::heaviside(input, values);

        ASSERT_TRUE(torch::allclose(result, expected, 1e-4));
    }

    // Test for zero input with custom h value
    TEST(HeavisideTest, ZeroInputCustomH) {
        custom_namespace::Heaviside heaviside_instance;
        torch::Tensor input = torch::tensor({2.0});
        torch::Tensor values = torch::tensor({0.7});
        
        // Call compute_heaviside with h = 0.7
        torch::Tensor result = heaviside_instance.compute_heaviside(input, values);

        // Expected output: tensor of 0.7 for input of zero
        torch::Tensor expected = torch::heaviside(input, values);

        ASSERT_TRUE(torch::allclose(result, expected, 1e-4));
    }

    // Test for negative inputs to compute_heaviside
    TEST(HeavisideTest, NegativeInput) {
        custom_namespace::Heaviside heaviside_instance;
        torch::Tensor input = torch::tensor({-1.0, 2.0, -3.0});
       torch::Tensor values = torch::tensor({-2.0, 4.0, 6.0});
        
        // Call compute_heaviside
        torch::Tensor result = heaviside_instance.compute_heaviside(input, values);

        // Expected output: tensor of zeros since input is negative
        torch::Tensor expected = torch::heaviside(input, values);

        ASSERT_TRUE(torch::allclose(result, expected, 1e-4));
    }

} // namespace custom_namespace

File: cerebras/test/test_heaviside.py
import torch, pytest
import custom_module

@pytest.mark.parametrize("input, values", [
    (torch.tensor([3.0, 0.0, -2.0]), torch.tensor([3.0, 0.0, 3.0])),   
    (torch.tensor([5.0]), torch.tensor([-3.0])), 
    (torch.tensor([1.5, -1.5]), torch.tensor([3.0, 2.0])),  
    (torch.tensor([0.0]), torch.tensor([0.0])),  
])
def test_heaviside(input, values):

    # Create an instance of Heaviside
    ch = custom_module.Heaviside()

    # Perform custom Heaviside
    custom_heaviside = ch.compute_heaviside(input, values)

    torch_heaviside = torch.heaviside(input, values)

    # Assert that the custom result matches the expected output
    assert torch.allclose(custom_heaviside, torch_heaviside, rtol=1e-05, atol=1e-05), \
    f"custom_heaviside({input}, {values}) != torch.heaviside({input}, {values})"